luaui/Widgets/unit_smart_select.lua

```diff
@@ -129,46 +133,44 @@ function widget:ViewResize()
 end
 
 function widget:SelectionChanged(sel)
-	local mousePressed = select(3, spGetMouseState())
+	-- widgets already received what we selected on previous pass, bypass this event
+	if finishedSelection then
+		WG['smartselect'].updateSelection = false
+		finishedSelection = false
+		return
+	end
 
 	-- Check if engine has just deselected via mouserelease on selectbox.
-	-- We want to preserve smartselect state above any engine modifiers.
-	if inSelection and not mousePressed then
+	-- We want to ignore engine passed selection and make sure we retain smartselect state
+	if inSelection and not select(3, spGetMouseState()) then -- left mouse button
 		inSelection = false
 
-		spSelectUnitArray(selectedUnits)
-
-		local _, ctrl = spGetModKeyState()
-
-		if ctrl and #sel == 0 then
-			WG['smartselect'].updateSelection = false    -- widgethandler uses this to ignore the engine mouserelease selection
+		if #sel == 0 and not select(2, spGetModKeyState()) then -- ctrl
+			-- if empty selection box and engine hardcoded deselect modifier is not
+			-- pressed, user is selected empty space
+			-- we must clear selection to disambiguate from our own deselect modifier
+			spSelectUnitArray({})
+		else
+			-- widgethandler uses this to ignore the engine mouserelease selection
+			-- we don't want to pass the engine selection to other widgets
+			WG['smartselect'].updateSelection = false
+			-- we also want to override back from engine selection to our selection
+			spSelectUnitArray(selectedUnits)
+			-- we bypass next selectionchanged event since we already know what we
+			-- selected, Update() constantly selects units inside selection box until
+			-- it finishes
+			finishedSelection = true
 		end
 
 		return
 	end
 
-	local equalSelection = #selectedUnits == #sel
-
-	if equalSelection then
-		for i = 1, #sel do
-			if selectedUnits[i] ~= sel[i] then
-				equalSelection = false
-				break
-			end
-		end
-	end
-
 	selectedUnits = sel
-	if spGetActiveCommand() == 0 then
```

Comment:
This check, and the code above to check for equality between event and widget selections, is not necessary anymore as long as our cache check on `Update()` is working properly.